### ISSUE TITLE
PHP 8.1 compatibility issues

### ISSUE THREAD

**Issue by therealgilles** [pods-framework/pods#6579]
### Description

PHP 8.1 reports the following deprecation warnings:

```PHP
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of Pods\Whatsit::offsetExists($offset) should either be compatible with ArrayAccess::offsetExists(mixed $offset): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 351
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of & Pods\Whatsit::offsetGet($offset) should either be compatible with ArrayAccess::offsetGet(mixed $offset): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 362
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of Pods\Whatsit::offsetSet($offset, $value) should either be compatible with ArrayAccess::offsetSet(mixed $offset, mixed $value): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 375
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of Pods\Whatsit::offsetUnset($offset) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 389
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of Pods\Whatsit::jsonSerialize() should either be compatible with JsonSerializable::jsonSerialize(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 285
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of Pods\Whatsit::current() should either be compatible with Iterator::current(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 306
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of Pods\Whatsit::next() should either be compatible with Iterator::next(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 322
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of Pods\Whatsit::key() should either be compatible with Iterator::key(): mixed, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 315
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of Pods\Whatsit::valid() should either be compatible with Iterator::valid(): bool, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 329
[08-Aug-2022 00:16:49 UTC] PHP Deprecated:  Return type of Pods\Whatsit::rewind() should either be compatible with Iterator::rewind(): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in .../public_html/wp-content/plugins/pods/src/Pods/Whatsit.php on line 299
```

### Version

2.8.23

**Comment by sc0ttkclark** [comment]
Will get this fixed in Pods 2.9.1

**Comment by sc0ttkclark** [comment]
Most of this can't be fixed until Pods 3.0 which will increase the minimum PHP version to 7.2+.

Since the minimum is currently PHP 5.6+, we don't have a way to use return types or the type hinting on the parameters because they won't match up.

We can't even use `#[\ReturnTypeWillChange]` because that's PHP 8.0+.

**Comment by sc0ttkclark** [comment]
Fixed via https://github.com/pods-framework/pods/commit/721418f0f97eb88e9ed5f98db668e81ef2e735bc

**Comment by ztackett11** [comment]
So PODS 3.0 will not have compatibility with PHP 8.0+?

Asking because PHP 7.4 is fully EoL at the end of this month. So having to stay at an PHP version that can even be patched seems like a pretty significant risk.

**Comment by sc0ttkclark** [comment]
@ztackett11 Pods 3.0 will have full PHP 8.1 support, we have to drop support for earlier PHP versions that cannot be easily supported alongside PHP 8.1.

Currently looking at a minimum PHP 8.0 requirement if we can manage that with our biggest project implementers.

**Comment by ztackett11** [comment]
Any idea on an expected release date for pods 3.0?

**Comment by sc0ttkclark** [comment]
We're still fundraising the remaining development costs but the current hold up is testing/QA before we go beta. I should get through that this weekend.

Once in beta, final release may be tricky with the holidays happening but I'll be doing my best to get it out in late November or early December.

Thanks for your patience, just hold tight a little longer.

**Comment by sc0ttkclark** [comment]
I've gone ahead and added tweaks for Pods 2.9.10 to bypass the deprecated messages for these methods (using the notice suggestion of the `#[\ReturnTypeWillChange]` attribute).

Pods 3.0 will implement the proper return types here. Reopening to make sure that process happens to the new ones identified in this batch of testing which I had found more.

**Comment by sc0ttkclark** [comment]
Fixed in Pods 3.0
